docs/design/2024-05-23-predicate-columns.md

```diff
@@ -0,0 +1,303 @@
+
+### Using Predicate Columns
+
+We can use the predicate columns in both automatic statistics collection and manual statistics collection.
+
+- Automatic statistics collection: When we trigger the automatic statistics collection, we can read the predicate columns from the system table and only analyze these columns.
+- Manual statistics collection: When we trigger the manual statistics collection, we can use the `ANALYZE TABLE tbl_name PREDICATE_COLUMNS` syntax to read the predicate columns from the system table and only analyze these columns.
+
+To comprehend the application of predicate columns in automatic statistics collection, it's essential to understand how TiDB persists the analyze options.
+
+In TiDB, we use a system table to store the analyze options, called `mysql.analyze_options`:
+
+```sql
+CREATE TABLE IF NOT EXISTS mysql.analyze_options (
+    table_id BIGINT(64) NOT NULL,
+    sample_num BIGINT(64) NOT NULL DEFAULT 0,
+    sample_rate DOUBLE NOT NULL DEFAULT -1,
+    buckets BIGINT(64) NOT NULL DEFAULT 0,
+    topn BIGINT(64) NOT NULL DEFAULT -1,
+    column_choice enum('DEFAULT','ALL','PREDICATE','LIST') NOT NULL DEFAULT 'DEFAULT',
+    column_ids TEXT(19372),
+    PRIMARY KEY (table_id) CLUSTERED
+);
+```
+
+We can focus on the `column_choice` column, which has three different column options in the analyze statement. The corresponding relations are as follows:
+
+| Analyze Statement                                | column_choice | column_ids | mysql.column_stats_usage                                                 | Explain                                                                  |
+|--------------------------------------------------|---------------|------------|--------------------------------------------------------------------------|--------------------------------------------------------------------------|
+| ANALYZE TABLE t;                                 | DEFAULT(ALL)  | None       | None                                                                     | It will analyze all analyzable columns from the table.                   |
+| ANALYZE TABLE t LIST COLUMNS col1, col2;         | LIST          | col1, col2 | None                                                                     | It will only analyze col1 and col2.                                      |
+| ANALYZE TABLE t PREDICATE FOR PREDICATE COLUMNS; | PREDICATE     | None       | All predicate columns were collected before in mysql.column_stats_usage. | It will only analyze columns that exist in the mysql.column_stats_usage. |
+
+As you can see, we pick PREDICATE as the column_choice for the ANALYZE TABLE t PREDICATE COLUMNS statement. At the same time, we now consider DEFAULT to be ALL, but to support predicate columns during auto-analyze, we need to change the definition of DEFAULT.
+
+| Predicate Column Feature Status | Predicate Columns in `sys.sql_column_stats_us` | Meaning                                                       |
```

Comment:
I would like to explain why this name was chosen. I used AI to help me generate the markdown table from the image (because Feishu doesn't support export as markdown). It seemed the AI was cheating on me and I fixed one but missed this one. Nice catch!